clients/iceberg-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/NessieUtils.scala

```diff
@@ -0,0 +1,205 @@
+      .build()
+  }
+
+  def setCurrentRef(
+      currentCatalog: CatalogPlugin,
+      catalog: Option[String],
+      ref: Reference
+  ): Reference = {
+    val catalogName = catalog.getOrElse(currentCatalog.name)
+    val catalogImpl =
+      SparkSession.active.sessionState.catalogManager.catalog(catalogName)
+    SparkSession.active.sparkContext.conf
+      .set(s"spark.sql.catalog.$catalogName.ref", ref.getName)
+    val catalogConf = SparkSession.active.sparkContext.conf
+      .getAllWithPrefix(s"spark.sql.catalog.$catalogName.")
+      .toMap
+      .asJava
+    catalogImpl.initialize(
+      catalogName,
+      new CaseInsensitiveStringMap(catalogConf)
+    )
+    getCurrentRef(currentCatalog, catalog)
+//    if (isIcebergNessie(catalogImpl)) {
+//      setIcebergRef(getIcebergSparkCatalog(catalogImpl), ref)
+//    } else if (isDeltaNessie) {
+//      // todo, how do we get current Ref?
+//      null
+//    } else {
+//      throw new AnalysisException("Catalog is not Nessie enabled")
+//    }
+  }
+
+  def getCurrentRef(
+      currentCatalog: CatalogPlugin,
+      catalog: Option[String]
+  ): Reference = {
+    val catalogName = catalog.getOrElse(currentCatalog.name)
+    val refName = SparkSession.active.sparkContext.conf
+      .get(s"spark.sql.catalog.$catalogName.ref")
+    nessieClient(currentCatalog, catalog).getTreeApi.getReferenceByName(refName)
+//    if (isIcebergNessie(catalogImpl)) {
+//      getIcebergRef(getIcebergSparkCatalog(catalogImpl))
+//    } else if (isDeltaNessie) {
+//      // todo, how do we get current Ref?
+//      null
+//    } else {
+//      throw new AnalysisException("Catalog is not Nessie enabled")
+//    }
+  }
+
+  private def isDeltaNessie: Boolean = {
+    SparkSession.active.sparkContext.conf
+      .contains("spark.delta.logStore.class") &&
+    SparkSession.active.sparkContext.conf
+      .get("spark.delta.logStore.class")
+      .contains("NessieLogStore")
+  }
+
+  private def isIcebergNessie(catalogImpl: CatalogPlugin): Boolean = {
+    catalogImpl.getClass.getCanonicalName.equals(
+      "org.apache.iceberg.spark.SparkCatalog"
+    ) || catalogImpl.getClass.getCanonicalName
+      .equals("org.apache.iceberg.spark.SparkSessionCatalog")
+  }
+
+  private def getIcebergSparkCatalog(
+      catalogImpl: CatalogPlugin
+  ): CatalogPlugin = {
+    val className = catalogImpl.getClass.getCanonicalName
+    className match {
+      case "org.apache.iceberg.spark.SparkSessionCatalog" =>
+        getIcebergCatalogFromSessionCatalog(catalogImpl)
+      case "org.apache.iceberg.spark.SparkCatalog" => catalogImpl
+      case _                                       => throw new AnalysisException("Unknown Iceberg Catalog")
+    }
+  }
+
+  private def getIcebergRef(catalogImpl: CatalogPlugin): Reference = {
+    val field = Class
+      .forName("org.apache.iceberg.spark.SparkCatalog")
+      .getDeclaredField("icebergCatalog")
+    field.setAccessible(true)
```

Comment:
Can we update Iceberg to get around this reflection hack?

Reply:
all reflection has been removed. Unused leftovers from previous iteration